src/main/java/com/snowflake/kafka/connector/records/RecordService.java

```diff
@@ -177,18 +211,56 @@ public Map<String, Object> getProcessedRecordForStreamingIngest(SinkRecord recor
     final Map<String, Object> streamingIngestRow = new HashMap<>();
     for (JsonNode node : row.content.getData()) {
       try {
-        streamingIngestRow.put(TABLE_COLUMN_CONTENT, MAPPER.writeValueAsString(node));
+        if (enableSchematization) {
+          Iterator<String> columnNames = node.fieldNames();
+          while (columnNames.hasNext()) {
+            String columnName = columnNames.next();
+            JsonNode columnNode = node.get(columnName);
+            columnName = formatColumnName(columnName);
+            Object columnValue;
+            if (columnNode.isArray()) {
+              List<String> itemList = new ArrayList<>();
+              ArrayNode arrayNode = (ArrayNode) columnNode;
+              Iterator<JsonNode> nodeIterator = arrayNode.iterator();
+              while (nodeIterator.hasNext()) {
+                JsonNode e = nodeIterator.next();
+                itemList.add(e.isTextual() ? e.textValue() : MAPPER.writeValueAsString(e));
+              }
+              columnValue = itemList;
+            } else if (columnNode.isTextual()) {
+              columnValue = columnNode.textValue();
+            } else {
+              columnValue = MAPPER.writeValueAsString(columnNode);
+            }
+
+            // while the value is always dumped into a string, the Streaming Ingest SDK
+            // will transformed the value according to its type in the table
+            // TODO: SNOW-630885 the record could come directly as a map, and don't have to be
+            //  dumped into a string but the original type could be used.
+            streamingIngestRow.put(columnName, columnValue);
+          }
+        } else {
+          streamingIngestRow.put(TABLE_COLUMN_CONTENT, MAPPER.writeValueAsString(node));
+        }
         if (metadataConfig.allFlag) {
           streamingIngestRow.put(TABLE_COLUMN_METADATA, MAPPER.writeValueAsString(row.metadata));
         }
-      } catch (JsonProcessingException e) {
+      } catch (IOException e) {
         // return an exception and propagate upwards
         e.printStackTrace();
       }
     }
+
     return streamingIngestRow;
   }
 
+  private String formatColumnName(String columnName) {
+    // the columnName has been checked and guaranteed not to be null or empty
+    return (columnName.charAt(0) == '"' && columnName.charAt(columnName.length() - 1) == '"')
+        ? columnName.substring(1, columnName.length() - 1)
+        : columnName.toUpperCase();
+  }
+
```

Comment:
can you write a blurb here mentioning what you do? may be in form of javadocs

Reply:
This is actually copied from the Ingest SDK to mimic the behavior. I will add the blurb.